File: backend/community/api/community.py
"""Community 公共 REST API（方案 §8，PR-B 只读 + PR-C 写纵切）。

前缀 /api/v1/community；读接口支持可选认证（D46），写接口必须登录。
游标规则（§8.2/D13）：公共列表/回复分页不绑定 principal；回复游标绑定
具体 post_id（D39）；通知游标绑定当前用户（私有游标）。
限流（§9.3/D41）：community.read 覆盖列表/详情/回复分页/通知读取；
写路径按 bucket 表（发帖小时级、回复分钟+小时双桶、点赞分钟）。
"""

from __future__ import annotations

from typing import Any
from uuid import UUID

from fastapi import APIRouter, Depends, Header, Query, Request
from pydantic import BaseModel, ConfigDict

from backend.auth.context import AuthContext
from backend.community.api.cursor import (
    issue_private_cursor,
    issue_public_cursor,
    resolve_private_cursor,
    resolve_public_cursor,
)
from backend.community.api.dependencies import (
    get_post_command_service,
    get_post_service,
    get_reply_service,
    rate_limit,
)
from backend.community.contracts.api import (
    BoardDetailResponse,
    BoardListResponse,
    CommunityNotification,
    CommunityNotificationPage,
    CommunityPostDetail,
    CommunityPostDetailResponse,
    CommunityPostSummary,
    CommunityReplyView,
    CreatePostRequest,
    CreateReplyRequest,
    PermissionsResponse,
    ResolveRequest,
)
from backend.community.contracts.errors import (
    CommunityContentInvalidError,
    CommunityCursorInvalidError,
)
from backend.community.services.post_service import PostReadService
from backend.settings import get_settings
from backend.shared.auth_context import get_auth_context, get_optional_auth_context

router = APIRouter(prefix="/api/v1/community", tags=["community"])


class PostListResponse(BaseModel):
    """帖子列表分页响应（§8.2/D45 信封）。"""

    model_config = ConfigDict(extra="forbid")

    items: list[CommunityPostSummary]
    next_cursor: str | None
    has_more: bool


#: 帖子列表游标 route 标识（§8.2 绑定项；与通知/回复路由互斥）
_POSTS_ROUTE = "community.posts"
#: 详情回复分页游标 route：绑定具体 post_id 防跨帖子复用（D39）
_REPLIES_ROUTE = "community.posts.detail.replies"


@router.get("/permissions", response_model=PermissionsResponse)
async def get_permissions(
    auth: AuthContext | None = Depends(get_optional_auth_context),
) -> PermissionsResponse:
    """当前用户是否社区管理员（§八 #21）。"""
    settings = get_settings()
    is_admin = auth is not None and auth.user_id in settings.community_admin_user_ids_set
    return PermissionsResponse(is_community_admin=is_admin)


@router.get("/boards", response_model=BoardListResponse)
async def list_boards(
    request: Request,
    auth: AuthContext | None = Depends(get_optional_auth_context),
    _rate: None = Depends(rate_limit("community.read")),
) -> BoardListResponse:
    """板块列表（§8.1）：只返回 status=active 板块。"""
    service = get_post_service(request)
    items = await service.list_boards()
    return BoardListResponse(items=items)


@router.get("/boards/{slug}", response_model=BoardDetailResponse)
async def get_board_detail(
    request: Request,
    slug: str,
    auth: AuthContext | None = Depends(get_optional_auth_context),
    _rate: None = Depends(rate_limit("community.read")),
) -> BoardDetailResponse:
    """板块详情（§八 #2）。"""
    service = get_post_service(request)
    return await service.get_board_detail_by_slug(slug, auth.user_id if auth else None)


@router.get("/posts", response_model=PostListResponse)
async def list_posts(
    request: Request,
    auth: AuthContext | None = Depends(get_optional_auth_context),
    board_id: UUID | None = Query(default=None),
    sort: str = Query(default="latest", pattern="^(latest|unanswered)$"),
    cursor: str | None = Query(default=None),
    limit: int = Query(default=20, ge=1, le=50),
    _rate: None = Depends(rate_limit("community.read")),
) -> PostListResponse:
    """帖子列表（§8.2）：latest/unanswered + 板块筛选 + keyset 游标。"""
    filters: dict[str, Any] = {"sort": sort}
    if board_id is not None:
        filters["board_id"] = str(board_id)
    payload = resolve_public_cursor(request, _POSTS_ROUTE, cursor, filters=filters)
    after_key: tuple[Any, ...] | None = None
    if payload is not None:
        sort_key = payload.get("sort_key")
        if not isinstance(sort_key, list) or len(sort_key) != 3:
            raise CommunityCursorInvalidError("游标缺少完整排序键")
        after_key = (bool(sort_key[0]), sort_key[1], UUID(str(sort_key[2])))
    service: PostReadService = get_post_service(request)
    items, next_key, has_more = await service.list_posts(
        viewer_user_id=auth.user_id if auth else None,
        board_id=board_id,
        sort=sort,
        after_key=after_key,
        limit=limit,
    )
    next_cursor = issue_public_cursor(
        request,
        route=_POSTS_ROUTE,
        filters=filters,
        next_after=next_key,
    )
    return PostListResponse(items=items, next_cursor=next_cursor, has_more=has_more)


@router.get("/posts/{post_id}", response_model=CommunityPostDetailResponse)
async def get_post_detail(
    request: Request,
    post_id: UUID,
    auth: AuthContext | None = Depends(get_optional_auth_context),
    reply_cursor: str | None = Query(default=None),
    reply_limit: int = Query(default=20, ge=1, le=50),
    _rate: None = Depends(rate_limit("community.read")),
) -> CommunityPostDetailResponse:
    """帖子详情 + 一页回复（§8.4）。"""
    route = f"{_REPLIES_ROUTE}:{post_id}"
    filters: dict[str, Any] = {}
    payload = resolve_public_cursor(request, route, reply_cursor, filters=filters)
    reply_after: tuple[Any, ...] | None = None
    if payload is not None:
        sort_key = payload.get("sort_key")
        if not isinstance(sort_key, list) or len(sort_key) != 2:
            raise CommunityCursorInvalidError("游标缺少完整排序键")
        reply_after = (sort_key[0], UUID(str(sort_key[1])))
    service: PostReadService = get_post_service(request)
    response, next_key = await service.get_post_detail(
        viewer_user_id=auth.user_id if auth else None,
        post_id=post_id,
        reply_after_key=reply_after,
        reply_limit=reply_limit,
    )
    if next_key is not None:
        response.replies.next_cursor = issue_public_cursor(
            request,
            route=route,
            filters=filters,
            next_after=next_key,
        )
    return response


# ---------------------------------------------------------------------------
# 写路径（§8.3–§8.5，PR-C）：幂等键必填、限流分桶、身份来自认证上下文
# ---------------------------------------------------------------------------


_IDEMPOTENCY_KEY_RE = r"^[\x21-\x7e]{1,200}$"


def _require_idempotency_key(idempotency_key: str | None) -> str:
    """§8.3：幂等键缺失/格式非法统一 422（§8.7 无专码，映射 CONTENT_INVALID）。"""
    if not idempotency_key:
        raise CommunityContentInvalidError(
            "缺少或非法的 Idempotency-Key（ASCII 可见字符，1–200）", field="Idempotency-Key"
        )
    return idempotency_key


@router.post("/posts", response_model=CommunityPostDetail, status_code=201)
async def create_post(
    request: Request,
    payload: CreatePostRequest,
    auth: AuthContext = Depends(get_auth_context),
    idempotency_key: str | None = Header(
        default=None, alias="Idempotency-Key", pattern=_IDEMPOTENCY_KEY_RE
    ),
    _rate: None = Depends(rate_limit("community.post.create")),
) -> CommunityPostDetail:
    """发帖（§8.3）：user_id 来自认证上下文；支持最多 3 张配图。"""
    _require_idempotency_key(idempotency_key)
    service = get_post_command_service(request)
    return await service.create_post(
        user_id=auth.user_id,
        board_id=payload.board_id,
        title=payload.title,
        body=payload.body,
        attachment_ids=payload.attachment_ids or [],
        idempotency_key=idempotency_key or "",
    )


@router.post("/posts/{post_id}/replies", response_model=CommunityReplyView, status_code=201)
async def create_reply(
    request: Request,
    post_id: UUID,
    payload: CreateReplyRequest,
    auth: AuthContext = Depends(get_auth_context),
    idempotency_key: str | None = Header(
        default=None, alias="Idempotency-Key", pattern=_IDEMPOTENCY_KEY_RE
    ),
    _rate: None = Depends(rate_limit("community.reply.create.minute")),
    _rate_hour: None = Depends(rate_limit("community.reply.create.hour")),
) -> CommunityReplyView:
    """回复（§8.4）：分钟 + 小时双窗口限流（§9.3）。"""
    _require_idempotency_key(idempotency_key)
    service = get_reply_service(request)
    row = await service.create_reply(
        user_id=auth.user_id,
        post_id=post_id,
        body=payload.body,
        idempotency_key=idempotency_key or "",
    )
    return PostReadService._to_reply(row, auth.user_id, solved=False)


@router.post("/posts/{post_id}/like", status_code=200)
async def like_post(
    request: Request,
    post_id: UUID,
    auth: AuthContext = Depends(get_auth_context),
    _rate: None = Depends(rate_limit("community.post.like")),
) -> dict[str, str]:
    """点赞（§8.5）：幂等；deleted/hidden → NOT_FOUND。"""
    await get_post_command_service(request).toggle_like(
        user_id=auth.user_id, post_id=post_id, like=True
    )
    return {"status": "ok"}


@router.delete("/posts/{post_id}/like", status_code=200)
async def unlike_post(
    request: Request,
    post_id: UUID,
    auth: AuthContext = Depends(get_auth_context),
    _rate: None = Depends(rate_limit("community.post.like")),
) -> dict[str, str]:
    """取消点赞（§8.5）：与点赞共用限流桶；幂等。"""
    await get_post_command_service(request).toggle_like(
        user_id=auth.user_id, post_id=post_id, like=False
    )
    return {"status": "ok"}


@router.post("/posts/{post_id}/resolve", status_code=200)
async def resolve_post(
    request: Request,
    post_id: UUID,
    payload: ResolveRequest,
    auth: AuthContext = Depends(get_auth_context),
) -> dict[str, str]:
    """标记解决/取消解决（§8.5：reply_id=null 表示取消）。"""
    await get_post_command_service(request).resolve(
        actor_user_id=auth.user_id, post_id=post_id, reply_id=payload.reply_id
    )
    return {"status": "ok"}


@router.delete("/posts/{post_id}", status_code=200)
async def delete_post(
    request: Request,
    post_id: UUID,
    auth: AuthContext = Depends(get_auth_context),
) -> dict[str, str]:
    """删除帖子（§11.1）：作者本人；重复删除幂等成功。"""
    await get_post_command_service(request).delete_post(actor_user_id=auth.user_id, post_id=post_id)
    return {"status": "ok"}


@router.delete("/posts/{post_id}/replies/{reply_id}", status_code=200)
async def delete_reply(
    request: Request,
    post_id: UUID,
    reply_id: UUID,
    auth: AuthContext = Depends(get_auth_context),
) -> dict[str, str]:
    """删除回复（§11.1）：作者本人；solved 回复清除解决标记（D34）。"""
    await get_reply_service(request).delete_reply(actor_user_id=auth.user_id, reply_id=reply_id)
    return {"status": "ok"}


# ---------------------------------------------------------------------------
# 社区通知（§8.6）：私有游标绑定当前用户（D13）
# ---------------------------------------------------------------------------

_NOTIFICATIONS_ROUTE = "community.notifications"


class ReadAllResponse(BaseModel):
    """两个 read-all 统一响应（§8.6 冻结：{"unread_count": 0}）。"""

    model_config = ConfigDict(extra="forbid")

    unread_count: int


@router.get("/notifications", response_model=CommunityNotificationPage)
async def list_notifications(
    request: Request,
    auth: AuthContext = Depends(get_auth_context),
    unread_only: bool = Query(default=False),
    cursor: str | None = Query(default=None),
    limit: int = Query(default=20, ge=1, le=50),
    _rate: None = Depends(rate_limit("community.read")),
) -> CommunityNotificationPage:
    """通知列表（§8.6）：只返回当前用户记录 + 全部未读数。"""
    from backend.community.api.dependencies import get_community_runtime
    from backend.community.persistence import notifications as notifications_repo

    filters: dict[str, Any] = {"unread_only": unread_only}
    payload = resolve_private_cursor(
        request,
        _NOTIFICATIONS_ROUTE,
        cursor,
        user_id=auth.user_id,
        filters=filters,
    )
    after: tuple[Any, UUID] | None = None
    if payload is not None:
        sort_key = payload.get("sort_key")
        if not isinstance(sort_key, list) or len(sort_key) != 2:
            raise CommunityCursorInvalidError("游标缺少完整排序键")
        after = (sort_key[0], UUID(str(sort_key[1])))
    session_factory = get_community_runtime(request).database.session_factory
    async with session_factory() as session:
        rows = await notifications_repo.list_notifications(
            session,
            user_id=auth.user_id,
            unread_only=unread_only,
            after=after,
            limit=limit,
        )
        unread = await notifications_repo.unread_count(session, user_id=auth.user_id)
    has_more = len(rows) > limit
    items = [
        CommunityNotification(
            notification_id=r["notification_id"],
            event_type=r["event_type"],
            title=r["title"],
            body=r["body"],
            read_at=r["read_at"],
            created_at=r["created_at"],
            post_id=r["post_id"],
            reply_id=r["reply_id"],
            board_slug=r["board_slug"],
        )
        for r in rows[:limit]
    ]
    next_cursor: str | None = None
    if has_more and rows:
        last = rows[limit - 1]
        next_cursor = issue_private_cursor(
            request,
            route=_NOTIFICATIONS_ROUTE,
            user_id=auth.user_id,
            filters=filters,
            next_after=(last["created_at"].isoformat(), str(last["notification_id"])),
        )
    return CommunityNotificationPage(
        items=items,
        next_cursor=next_cursor,
        has_more=has_more,
        unread_count=unread,
    )


@router.post("/notifications/{notification_id}/read", response_model=ReadAllResponse)
async def mark_notification_read(
    request: Request,
    notification_id: UUID,
    auth: AuthContext = Depends(get_auth_context),
) -> ReadAllResponse:
    """标记单条已读（§8.6：幂等；recipient 条件保证不能跨用户标记）。"""
    from backend.community.api.dependencies import get_community_runtime
    from backend.community.persistence import notifications as notifications_repo

    session_factory = get_community_runtime(request).database.session_factory
    async with session_factory() as session:
        async with session.begin():
            await notifications_repo.mark_read(
                session, user_id=auth.user_id, notification_id=notification_id
            )
            unread = await notifications_repo.unread_count(session, user_id=auth.user_id)
    return ReadAllResponse(unread_count=unread)


@router.post("/notifications/read-all", response_model=ReadAllResponse)
async def read_all_notifications(
    request: Request,
    auth: AuthContext = Depends(get_auth_context),
) -> ReadAllResponse:
    """全部已读（§8.6/D14：只更新当前认证用户的未读记录）。"""
    from backend.community.api.dependencies import get_community_runtime
    from backend.community.persistence import notifications as notifications_repo

    session_factory = get_community_runtime(request).database.session_factory
    async with session_factory() as session:
        async with session.begin():
            await notifications_repo.mark_all_read(session, user_id=auth.user_id)
            unread = await notifications_repo.unread_count(session, user_id=auth.user_id)
    return ReadAllResponse(unread_count=unread)

File: frontend/src/api/community.ts
// Community API 客户端（方案 §6.5 / §6.6，v1.6 冻结）。
// 经由共享请求层 client.ts 发出（挂 Bearer、带 credentials、401 刷新、
// PublicError 信封与 idempotencyKey）；前端不提交 user_id（§9.1）。

import { idempotencyKey, request } from "./client";

// ---------------------------------------------------------------------------
// 契约镜像类型（后端 backend/community/contracts/api.py，§6.6 冻结）
// ---------------------------------------------------------------------------

export interface CommunityBoard {
  board_id: string;
  slug: string;
  name: string;
  description: string;
  /** 0002 新增，列表恒返回；旧 fixture 可能缺省，视图层用 ?? 0 兜底 */
  post_count?: number;
  /** 0002 新增，列表恒返回；仅用于排序展示 */
  sort_order?: number;
}

/** 板块详情（§八 #2）：viewer_is_owner 恒 bool，匿名 false */
export interface CommunityBoardDetail {
  board_id: string;
  slug: string;
  name: string;
  description: string;
  post_count: number;
  created_at: string;
  viewer_is_owner: boolean;
}

/** §八 #2：板块详情为平铺对象（与后端 BoardDetailResponse 一致，无包裹层）。 */
export type CommunityBoardDetailResponse = CommunityBoardDetail;

/** 附件视图（D37 恒返回，按 position ASC；无附件为 []） */
export interface CommunityAttachment {
  attachment_id: string;
  url: string;
  width: number;
  height: number;
  mime: string;
  position: number;
}

/** POST /uploads 成功响应（201） */
export interface CommunityAttachmentUpload {
  attachment_id: string;
  url: string;
  mime: string;
  width: number;
  height: number;
  size_bytes: number;
}

export interface CommunityAuthor {
  display_name: string;
}

export interface CommunityPostSummary {
  post_id: string;
  board: CommunityBoard;
  author: CommunityAuthor;
  title: string; // Summary 恒非 null（列表只含 active）
  pinned: boolean;
  solved: boolean;
  reply_count: number;
  like_count: number;
  viewer_liked: boolean;
  created_at: string;
  last_activity_at: string;
  /** 0002 新增，恒返回数组；旧 fixture 可能缺省 */
  attachments?: CommunityAttachment[];
}

// §6.6：Detail 的 title 可空（deleted 墓碑）；Summary 恒非 null，
// 故不用 extends，显式组合全部字段
export interface CommunityPostDetail {
  post_id: string;
  board: CommunityBoard;
  author: CommunityAuthor;
  title: string | null; // deleted 时为 null（墓碑）
  pinned: boolean;
  solved: boolean;
  reply_count: number;
  like_count: number;
  viewer_liked: boolean;
  created_at: string;
  last_activity_at: string;
  body: string | null; // deleted 时为 null（墓碑）
  deleted: boolean;
  discussion_status: "open" | "closed";
  viewer_is_author: boolean;
  solved_reply_id: string | null;
  deleted_at: string | null;
  /** 0002 新增，恒返回数组；旧 fixture 可能缺省 */
  attachments?: CommunityAttachment[];
}

export interface CommunityReplyView {
  reply_id: string;
  author: CommunityAuthor;
  body: string | null; // deleted 时为 null（墓碑占位行）
  deleted: boolean;
  viewer_is_author: boolean;
  solved: boolean;
  created_at: string;
}

export interface CommunityNotification {
  notification_id: string;
  event_type:
    | "post_replied"
    | "reply_marked_solved"
    | "application_approved"
    | "application_rejected";
  title: string;
  body: string;
  read_at: string | null;
  created_at: string;
  post_id: string | null;
  reply_id: string | null;
  /** 0002 新增：四类通知恒非空；审核通知 post_id/reply_id 为 null */
  board_slug?: string | null;
}

// §19.45 Page 信封：{items, next_cursor, has_more}；通知页额外顶层 unread_count
export interface CursorPage<T> {
  items: T[];
  next_cursor: string | null;
  has_more: boolean;
}

export type CommunityPostListPage = CursorPage<CommunityPostSummary>;

export interface CommunityPostDetailResponse {
  post: CommunityPostDetail;
  replies: CursorPage<CommunityReplyView>;
}

export interface CommunityNotificationPage extends CursorPage<CommunityNotification> {
  unread_count: number;
}

// ---------------------------------------------------------------------------
// 只读接口（PR-B；写接口在 PR-C 追加）
// ---------------------------------------------------------------------------

/** 板块列表（§8.1）：只返回 active 板块。 */
export async function listBoards(): Promise<CommunityBoard[]> {
  const body = await request<{ items: CommunityBoard[] }>("GET", "/community/boards");
  return body.items;
}

/** 帖子列表（§8.2）：latest/unanswered + 板块筛选 + 游标分页。 */
export async function listPosts(params: {
  board_id?: string;
  sort?: "latest" | "unanswered";
  cursor?: string;
  limit?: number;
}): Promise<CommunityPostListPage> {
  const search = new URLSearchParams();
  if (params.board_id) search.set("board_id", params.board_id);
  if (params.sort && params.sort !== "latest") search.set("sort", params.sort);
  if (params.cursor) search.set("cursor", params.cursor);
  if (params.limit) search.set("limit", String(params.limit));
  const qs = search.toString();
  return request<CommunityPostListPage>(
    "GET",
    `/community/posts${qs ? `?${qs}` : ""}`,
  );
}

/** 帖子详情 + 一页回复（§8.4）。 */
export async function getPostDetail(params: {
  post_id: string;
  reply_cursor?: string;
  reply_limit?: number;
}): Promise<CommunityPostDetailResponse> {
  const search = new URLSearchParams();
  if (params.reply_cursor) search.set("reply_cursor", params.reply_cursor);
  if (params.reply_limit) search.set("reply_limit", String(params.reply_limit));
  const qs = search.toString();
  return request<CommunityPostDetailResponse>(
    "GET",
    `/community/posts/${params.post_id}${qs ? `?${qs}` : ""}`,
  );
}

// ---------------------------------------------------------------------------
// 写接口（§8.3–§8.5，PR-C）：均携带 Idempotency-Key（幂等重试不重复创建）
// ---------------------------------------------------------------------------

/** 发帖（§8.3）：user_id 由服务端认证上下文取得；attachment_ids 顺序即展示顺序。 */
export async function createPost(payload: {
  board_id: string;
  title: string;
  body: string;
  attachment_ids?: string[];
}): Promise<CommunityPostDetail> {
  return request<CommunityPostDetail>("POST", "/community/posts", {
    body: payload,
    idempotencyKey: idempotencyKey(),
  });
}

/** 回复（§8.4）。 */
export async function createReply(
  post_id: string,
  body: string,
): Promise<CommunityReplyView> {
  return request<CommunityReplyView>("POST", `/community/posts/${post_id}/replies`, {
    body: { body },
    idempotencyKey: idempotencyKey(),
  });
}

/** 点赞（§8.5：幂等）。 */
export async function likePost(post_id: string): Promise<void> {
  await request("POST", `/community/posts/${post_id}/like`);
}

/** 取消点赞（§8.5：幂等）。 */
export async function unlikePost(post_id: string): Promise<void> {
  await request("DELETE", `/community/posts/${post_id}/like`);
}

/** 标记解决/取消解决（§8.5：reply_id=null 表示取消解决）。 */
export async function resolvePost(post_id: string, reply_id: string | null): Promise<void> {
  await request("POST", `/community/posts/${post_id}/resolve`, {
    body: { reply_id },
  });
}

/** 删除帖子（§11.1：作者本人；重复删除幂等成功）。 */
export async function deletePost(post_id: string): Promise<void> {
  await request("DELETE", `/community/posts/${post_id}`);
}

/** 删除回复（§11.1）。 */
export async function deleteReply(post_id: string, reply_id: string): Promise<void> {
  await request("DELETE", `/community/posts/${post_id}/replies/${reply_id}`);
}

// ---------------------------------------------------------------------------
// 社区通知（§8.6）
// ---------------------------------------------------------------------------

/** 通知列表：只返回当前用户记录 + 全部未读数。 */
export async function listCommunityNotifications(params: {
  unread_only?: boolean;
  cursor?: string;
  limit?: number;
} = {}): Promise<CommunityNotificationPage> {
  const search = new URLSearchParams();
  if (params.unread_only) search.set("unread_only", "true");
  if (params.cursor) search.set("cursor", params.cursor);
  if (params.limit) search.set("limit", String(params.limit));
  const qs = search.toString();
  return request<CommunityNotificationPage>(
    "GET",
    `/community/notifications${qs ? `?${qs}` : ""}`,
  );
}

/** 全部已读（§8.6）：只更新当前认证用户的未读记录。 */
export async function markAllCommunityNotificationsRead(): Promise<{
  unread_count: number;
}> {
  return request<{ unread_count: number }>(
    "POST",
    "/community/notifications/read-all",
  );
}

// ---------------------------------------------------------------------------
// 社区重建新增接口（docs/community-rebuild-plan.md §八 #2/#12–#18，v3.9 冻结）
// 注意：这些路由仅在后端 COMMUNITY_V2_ENABLED=true 时挂载
// ---------------------------------------------------------------------------

export type CommunityBoardApplicationStatus = "pending" | "approved" | "rejected";

/** 申请对象（D44：不返回 reviewer_id；mine 与 admin 列表同一形状） */
export interface CommunityBoardApplication {
  application_id: string;
  name: string;
  slug: string;
  description: string;
  reason: string;
  status: CommunityBoardApplicationStatus;
  board_id: string | null;
  reviewed_at: string | null;
  reject_reason: string | null;
  created_at: string;
}

export type CommunityBoardApplicationPage = CursorPage<CommunityBoardApplication>;

export interface CommunityPermissions {
  is_community_admin: boolean;
}

/** 板块详情（§八 #2）。 */
export async function getBoardDetail(slug: string): Promise<CommunityBoardDetailResponse> {
  return request<CommunityBoardDetailResponse>("GET", `/community/boards/${slug}`);
}

/** 当前用户社区权限（§八 #18）：管理员入口展示依据（D23）。 */
export async function getCommunityPermissions(): Promise<CommunityPermissions> {
  return request<CommunityPermissions>("GET", "/community/permissions");
}

/** 图片上传（§八 #12）：multipart 字段 file + Idempotency-Key；支持 AbortController 取消。 */
export async function uploadAttachment(
  file: File,
  signal?: AbortSignal,
): Promise<CommunityAttachmentUpload> {
  const formData = new FormData();
  formData.append("file", file);
  return request<CommunityAttachmentUpload>("POST", "/community/uploads", {
    formData,
    idempotencyKey: idempotencyKey(),
    signal,
  });
}

/** 提交建吧申请（§八 #13）：携带 Idempotency-Key。 */
export async function createBoardApplication(payload: {
  name: string;
  slug: string;
  description: string;
  reason: string;
}): Promise<CommunityBoardApplication> {
  return request<CommunityBoardApplication>("POST", "/community/applications", {
    body: payload,
    idempotencyKey: idempotencyKey(),
  });
}

/** 我的申请列表（§八 #14）：created_at DESC 游标分页。 */
export async function listMyBoardApplications(params: {
  cursor?: string;
  limit?: number;
} = {}): Promise<CommunityBoardApplicationPage> {
  const search = new URLSearchParams();
  if (params.cursor) search.set("cursor", params.cursor);
  if (params.limit) search.set("limit", String(params.limit));
  const qs = search.toString();
  return request<CommunityBoardApplicationPage>(
    "GET",
    `/community/applications/mine${qs ? `?${qs}` : ""}`,
  );
}

/** 管理员审核列表（§八 #15）：status ∈ pending/approved/rejected/all，created_at ASC。 */
export async function listAdminBoardApplications(params: {
  status?: "pending" | "approved" | "rejected" | "all";
  cursor?: string;
  limit?: number;
} = {}): Promise<CommunityBoardApplicationPage> {
  const search = new URLSearchParams();
  // 后端缺省 status=None 视为 all；审核台默认 pending 必须显式传
  search.set("status", params.status ?? "pending");
  if (params.cursor) search.set("cursor", params.cursor);
  if (params.limit) search.set("limit", String(params.limit));
  const qs = search.toString();
  return request<CommunityBoardApplicationPage>(
    "GET",
    `/community/admin/applications${qs ? `?${qs}` : ""}`,
  );
}

/** 审核通过（§八 #16）：无 Idempotency-Key（D38 行锁 + 状态门）。 */
export async function approveBoardApplication(
  applicationId: string,
): Promise<CommunityBoardApplication> {
  return request<CommunityBoardApplication>(
    "POST",
    `/community/admin/applications/${applicationId}/approve`,
  );
}

/** 审核拒绝（§八 #17）：reason 1–200 字符。 */
export async function rejectBoardApplication(
  applicationId: string,
  reason: string,
): Promise<CommunityBoardApplication> {
  return request<CommunityBoardApplication>(
    "POST",
    `/community/admin/applications/${applicationId}/reject`,
    { body: { reason } },
  );
}
